src/client/visualizations/heat-map/heatmap-rectangles.tsx

```diff
@@ -0,0 +1,192 @@
+ * Copyright 2017-2018 Allegro.pl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { HeatmapRect } from "@vx/heatmap";
+import { scaleLinear } from "@vx/scale";
+import { Dataset, Datum } from "plywood";
+import * as React from "react";
+import { GlobalEventListener } from "../../components/global-event-listener/global-event-listener";
+import { SPLIT } from "../../config/constants";
+import { HeatMapRectangleRow } from "./heatmap-rectangle-row";
+
+const white = "#fff";
+const orange = "#ff5a00";
+
+const max = (data: any[], datumToNumber: (d: any) => number) =>
+  Math.max(...data.map(datumToNumber));
+const min = (data: any[], datumToNumber: (d: any) => number) =>
+  Math.min(...data.map(datumToNumber));
+
+const bins = (d: Datum) => (d[SPLIT] as Dataset).data;
+
+export interface RectangleData {
+  xLabel: string;
+  yLabel: string;
+  datum: Datum;
+  x: number;
+  y: number;
+  column: number;
+  row: number;
+}
+
+interface HeatMapRectanglesProps {
+  dataset: Datum[];
+  tileSize?: number;
+  measureName: string;
+  leftLabelName: string;
+  topLabelName: string;
+  hoveredRectangle?: RectangleData;
+  onHover?: (data: RectangleData) => void;
+  onHoverStop?: () => void;
+}
+
+export class HeatMapRectangles extends React.Component<HeatMapRectanglesProps> {
+  private rect: HTMLDivElement | null = null;
+
+  handleMouseMove = (event: MouseEvent) => {
+    const { clientX: x, clientY: y } = event;
+    const {
+      onHoverStop = () => {},
+      onHover = () => {},
+      leftLabelName,
+      topLabelName,
+      dataset
+    } = this.props;
+
+    const {
+      xScale,
+      yScale
+    } = this.setup();
+
+    if (!this.rect) {
+      return;
+    }
+    const { top, bottom, left, right } = this.rect.getBoundingClientRect();
+
+    if ((y < top || y > bottom) || (x < left || x > right)) {
+      onHoverStop();
+      return;
+    }
+
+    const xPosition = Math.floor(xScale.invert(x - left));
+    const yPosition = Math.floor(yScale.invert(y - top));
+
+    const hoveredBins = dataset[yPosition];
+    if (!hoveredBins) {
+      return;
+    }
+
+    const hoveredBin = bins(hoveredBins)[xPosition];
+    if (!hoveredBin) {
+      return;
+    }
+
+    onHover({
+      datum: hoveredBin,
+      xLabel: hoveredBins[leftLabelName] as string,
+      yLabel: hoveredBin[topLabelName] as string,
+      x,
+      y,
+      column: yPosition,
+      row: xPosition
+    });
+  }
+
+  private setup() {
```

Comment:
We could win some performance with smart memoization of this code. Each render we're running through whole dataset and thrash scales...